fix(Core/Commands): Waypoint show command on/off/first/last #22740

SAS2000 opened the pull request:
NOTE: This PR is a direct update on my previous issue: https://github.com/azerothcore/azerothcore-wotlk/pull/22727
<!-- First of all, THANK YOU for your contribution. -->

## Changes Proposed:
<!-- If your pull request promotes complex changes that require a detailed explanation, please describe them in detail specifying what your solution is and what is it meant to address. -->
This PR proposes changes to:
**-  [ ] Core (units, players, creatures, game systems).**


## Issues Addressed:
<!-- If your fix has a relating issue, link it below -->
- Closes #22740 



## Tests Performed:
<!-- Does it build without errors? Did you test in-game? What did you test? On which OS did you test? Describe any other tests performed -->
This PR has been:
**- [ ] Tested in-game by the author.**


## How to Test the Changes:
<!-- Describe in a detailed step-by-step order how to test the changes -->

**- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue**

## Known Issues and TODO List:
<!-- Is there anything else left to do after this PR? -->

- The wp show on/first/last implementation still not perfect. If we would like make it more robust we should also save the map_id and not just the dbGuid to the waypoint table. That way we can delete all waypoints in one go not just the ones that are in the same map as we are. But because it is a gm tool i think the proposed implementation is good enough.

<!-- If you intend to contribute repeatedly to our project, it is a good idea to join our discord channel. We set ranks for our contributors and give them access to special resources or knowledge: https://discord.com/invite/GyFvXpk7)
     Do not remove the instructions below about testing, they will help users to test your PR -->
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

SAS2000 committed:
fix(Core/Commands): Waypoint show on/off bug

Current implementation cannot delete waypoints

Closes AzerothCore issue #22707

On branch fix-issue-22726
  Changes to be committed:
    modified:   src/server/scripts/Commands/cs_wp.cpp

SAS2000 committed:
Merge branch 'master' into fix-issue-22726

SAS2000 committed:
On branch fix-issue-22726
  Changes to be committed:
    modified:   src/server/scripts/Commands/cs_wp.cpp

SAS2000 committed:
Merge branch 'master' into fix-issue-22726

SAS2000 committed:
On branch fix-issue-22726
  Changes to be committed:
    modified:   src/server/scripts/Commands/cs_wp.cpp

SAS2000 committed:
Merge branch 'master' into fix-issue-22726

SAS2000 committed:
fix(Core/Commands): wp show on/off/first/last

wp show first/last not works as expected

Closes AzerothCore issue #22740

On branch fix-issue-22740
  Changes to be committed:
	new file:   data/sql/updates/db_world/2025_08_28_00.sql
	modified:   src/server/database/Database/Implementation/WorldDatabase.cpp
	modified:   src/server/database/Database/Implementation/WorldDatabase.h
	modified:   src/server/game/Miscellaneous/Language.h
	modified:   src/server/scripts/Commands/cs_wp.cpp

SAS2000 committed:
On branch fix-issue-22740
  Changes to be committed:
	modified:   src/server/scripts/Commands/cs_wp.cpp

TheSCREWEDSoftware commented:
Does this PR require https://github.com/azerothcore/azerothcore-wotlk/pull/22727 to work?

SAS2000 committed:
Merge branch 'master' into fix-issue-22740

SAS2000 committed:
Merge remote-tracking branch 'origin/master' into fix-issue-22740

Conflicts:
	data/sql/updates/db_world/2025_08_28_00.sql

Changes to be committed:
	modified:   data/sql/updates/db_world/2025_08_28_00.sql
	new file:   data/sql/updates/db_world/2025_08_28_01.sql
	new file:   data/sql/updates/db_world/2025_08_29_00.sql
	modified:   src/server/scripts/World/scourge_invasion.cpp

SAS2000 commented:
> Does this PR require #22727 to work?

No, but I used that branch as a base. All the issues mentioned there (and how can you tested the fixes) are still valid here but I added some new improvements.

Only this code should be reviewed and merged (other one just stay there for reference)

SAS2000 commented on data/sql/updates/db_world/2025_08_28_01.sql:
```diff
@@ -0,0 +1,5 @@
+-- add wp show first/last to command help
+UPDATE `command` SET
+`help`='Syntax: .wp show $option\nOptions:\non $pathid (or selected creature with loaded path) - Show path\nfirst $pathid (or selected creature with loaded path) - Show first waypoint\nlast $pathid (or selected creature with loaded path) - Show last waypoint\noff - Hide path (all waypoints in current map)\ninfo $selected_waypoint - Show info for selected waypoint.'
+WHERE `name` = 'wp show';
```
Originally the first/last option was missing from the description

SAS2000 commented on src/server/database/Database/Implementation/WorldDatabase.cpp:
```diff
@@ -53,12 +53,12 @@ void WorldDatabaseConnection::DoPrepareStatements()
-    PrepareStatement(WORLD_SEL_WAYPOINT_DATA_POS_LAST_BY_ID, "SELECT position_x, position_y, position_z, orientation FROM waypoint_data WHERE id = ? ORDER BY point DESC LIMIT 1", CONNECTION_SYNCH);
+    PrepareStatement(WORLD_SEL_WAYPOINT_DATA_POS_BY_ID, "SELECT point, position_x, position_y, position_z, orientation, wpguid FROM waypoint_data WHERE id = ?", CONNECTION_SYNCH);
+    PrepareStatement(WORLD_SEL_WAYPOINT_DATA_POS_FIRST_BY_ID, "SELECT point, position_x, position_y, position_z, orientation, wpguid FROM waypoint_data WHERE id = ? ORDER BY point ASC LIMIT 1", CONNECTION_SYNCH);
+    PrepareStatement(WORLD_SEL_WAYPOINT_DATA_POS_LAST_BY_ID, "SELECT point, position_x, position_y, position_z, orientation, wpguid FROM waypoint_data WHERE id = ? ORDER BY point DESC LIMIT 1", CONNECTION_SYNCH);
```
First and last was different. First used point=1 which can be error prone if the first point was deleted. Also query the point number is important so we can show it in the UI as level.
I unified these three queries so we can use the same loop to parse the result.

SAS2000 commented on src/server/game/Miscellaneous/Language.h:
```diff
@@ -287,8 +287,8 @@ enum AcoreStrings
-    LANG_WAYPOINT_TOOFAR2               = 244,
-    LANG_WAYPOINT_TOOFAR3               = 245,
+    LANG_WAYPOINT_TOOFAR2               = 244,  // Not used
+    LANG_WAYPOINT_TOOFAR3               = 245,  // Not used
```
I wasn't sure how to handle this. Because I didn't want to delte them I marked them as unused.

SAS2000 committed:
On branch fix-issue-22740
  Changes to be committed:
    modified:   src/server/scripts/Commands/cs_wp.cpp

SAS2000 committed:
Merge branch 'master' into fix-issue-22740

SAS2000 committed:
Merge branch 'master' into fix-issue-22740

SAS2000 committed:
On branch fix-issue-22740
Changes to be committed:
	renamed:    data/sql/updates/db_world/2025_08_28_01.sql -> data/sql/updates/db_world/2025_10_12_07.sql

SAS2000 committed:
Merge branch 'master' into fix-issue-22740

SAS2000 committed:
Merge branch 'master' into fix-issue-22740

sudlud committed:
Merge branch 'master' into fix-issue-22740

TheSCREWEDSoftware committed:
Merge branch 'master' into fix-issue-22740

SAS2000 committed:
Merge branch 'master' into fix-issue-22740

SAS2000 committed:
Merge branch 'master' into fix-issue-22740

sudlud committed:
Merge branch 'master' into fix-issue-22740

sudlud committed:
Merge branch 'master' into fix-issue-22740

sudlud reviewed (commented):
Hi @SAS2000, thank you for this contribution and for the detailed issue write-up that goes with it. I am sorry it took so long to get a proper second look at this. The refactor is a clear improvement: unifying `on`/`first`/`last` behind a single `AddWaypointsByPreparedStatement` helper, making `first`/`last` DB-backed and info-able, switching `first` to `ORDER BY point ASC LIMIT 1` so it survives deletion of point 1, and scoping `wp show off` to the current map are all the right calls and they match what #22740 asked for.

There are a few things that need to be sorted before this can be merged, mostly mechanical, plus a few cleanups.

## Required before merge

1. **Rebase onto current master (the branch is conflicting).** The branch is based on an older snapshot, and two things have moved since:
   - **RBAC permissions.** The command table here registers the `wp` subcommands with `SEC_ADMINISTRATOR`, but master has migrated them to `rbac::RBAC_PERM_COMMAND_WP_*`. As-is, merging would revert that. Please keep the `rbac::RBAC_PERM_COMMAND_WP_SHOW` (and siblings) when you rebase.

2. **Move the SQL update into `pending_db_world/`.** The PR adds `data/sql/updates/db_world/2025_10_12_07.sql` directly into the merged directory, with a hand-written `-- DB update 2025_10_12_06 -> 2025_10_12_07` header. That directory is import-only, and the date is now well behind master, which breaks the update chain and is part of why the branch conflicts. Please:
   - delete `db_world/2025_10_12_07.sql`,
   - run `./create_sql.sh` inside `data/sql/updates/pending_db_world/` and put the same `UPDATE command ...` statement into the generated `rev_<timestamp>.sql`.

   The import step assigns the correct date and chains the header at merge time. The SQL content itself is fine.

3. **CodeFactor (the failing check you asked about).** It is flagging style. A couple of spots to check against `apps/codestyle/codestyle-cpp.py`:
   - `namespace {` and the lambda in `RemoveWaypointById` use K&R braces; the project uses Allman.
   - `const ObjectGuid::LowType guid` should be `ObjectGuid::LowType const guid` (the `Type const` rule).

## Cleanups requested in the Copilot review (all valid)

4. **Remove the `DEBUG:` chat messages.** `wp on`/`first`/`last`/`info` and the final fallthrough all print `DEBUG:` lines to the GM's chat. Some predate this PR, but since this rewrites the function it is a good time to drop them (or route to logs). While there, the final `"no valid command found"` branch returns `true`; returning `false` would surface the command help on a bad subcommand.

5. **Hoist the `waypoint_data` update out of the loop in `RemoveWaypointById`.** When `deleteFromWaypointData` is true, `WORLD_UPD_WAYPOINT_DATA_BY_WPGUID` runs once per entry in the `equal_range`, even though it is identical each time. It only needs to run once.

6. **Drop the redundant `if (result)` in the `off` branch.** It is immediately preceded by `if (!result) return false;`, so the inner check is dead.

## Edge cases worth a note (non-blocking, your call)

7. **Orphan rows when replacing a visual on another map.** In the replace path, `RemoveWaypointById(handler, wpguid)` is called with `deleteFromWaypointData = false`. If that old visual creature is not in the current map, the call returns `false`, the old creature row is left in the DB, and then `wpguid` is overwritten with the new spawn, leaking the old row. This is the tension you already noted in your TODO about storing `map_id`, so I am fine treating it as a known limitation rather than a blocker. A short comment in the code acknowledging it would be enough.

8. **`LoadCreatureFromDB` failure path.** If `SaveToDB` succeeds but `LoadCreatureFromDB` fails, the saved creature row and the `wpguid` reference are left behind. This is extremely unlikely (it reloads the row just written) and already exists in master, so I am not asking you to fix it here, just flagging it.

9. Minor: the `Optional<ObjectGuid::LowType> guid` parameter is actually a path id. Renaming it to `pathId` would read better.

Thanks again for sticking with this and for the careful issue analysis. Once the rebase, the SQL relocation, and the CodeFactor and DEBUG cleanups are done, this should be in good shape.
